ext/KrylovKitChainRulesCoreExt/eigsolve.jl

```diff
@@ -0,0 +1,322 @@
+            ws[i] = scale(v, Δλ)
+            continue
+        end
+
+        # General case :
+
+        # for the case where `f` is a real matrix, we can expect the following simplication
+        # TODO: can we implement this within our general approach, or generalise this to also
+        # cover the case where `f` is a function?
+        # if i > 1 && eltype(A) <: Real &&
+        #    vals[i] == conj(vals[i - 1]) && Δvals[i] == conj(Δvals[i - 1]) &&
+        #    vecs[i] == conj(vecs[i - 1]) && Δvecs[i] == conj(Δvecs[i - 1])
+        #     ws[i] = conj(ws[i - 1])
+        #     continue
+        # end
+
+        if isa(Δv, AbstractZero)
+            b = (zerovector(v), convert(T, Δλ))
+        else
+            vdΔv = inner(v, Δv)
+            gaugeᵢ = abs(imag(vdΔv))
+            if gaugeᵢ > alg_primal.tol && alg_rrule.verbosity >= 1
+                @warn "`eigsolve` cotangent for eigenvector $i is sensitive to gauge choice: (|gaugeᵢ| = $gaugeᵢ)"
+            end
+            Δv = add(Δv, v, -vdΔv)
+            b = (Δv, convert(T, Δλ))
+        end
+        w, reverse_info = let λ = λ, v = v
+            linsolve(b, zerovector(b), alg_rrule) do x
+                x1, x2 = x
+                y1 = VectorInterface.add!!(VectorInterface.add!!(fᴴ(x1), x1, conj(λ), -1),
+                                           v, x2)
+                y2 = inner(v, x1)
+                return (y1, y2)
+            end
+        end
+        if info.converged >= i && reverse_info.converged == 0 && alg_rrule.verbosity >= 0
+            @warn "`eigsolve` cotangent linear problem ($i) did not converge, whereas the primal eigenvalue problem did: normres = $(reverse_info.normres)"
+        elseif abs(w[2]) > alg_rrule.tol && alg_rrule.verbosity >= 0
+            @warn "`eigsolve` cotangent linear problem ($i) returns unexpected result: error = $(w[2])"
+        end
+        ws[i] = w[1]
+    end
+    return ws
+end
+
+function compute_eigsolve_pullback_data(Δvals, Δvecs, vals, vecs, info, which, fᴴ, T,
+                                        alg_primal::Arnoldi, alg_rrule::Arnoldi)
+    n = length(Δvecs)
+    G = zeros(T, n, n)
+    VdΔV = zeros(T, n, n)
+    for j in 1:n
+        for i in 1:n
+            if i < j
+                G[i, j] = conj(G[j, i])
+            elseif i == j
+                G[i, i] = norm(vecs[i])^2
+            else
+                G[i, j] = inner(vecs[i], vecs[j])
+            end
+            if !(Δvecs[j] isa AbstractZero)
+                VdΔV[i, j] = inner(vecs[i], Δvecs[j])
+            end
+        end
+    end
+
+    # components along subspace spanned by current eigenvectors
+    tol = alg_primal.tol
+    mask = abs.(transpose(vals) .- vals) .< tol
+    gaugepart = VdΔV[mask] - Diagonal(real(diag(VdΔV)))[mask]
+    Δgauge = norm(gaugepart, Inf)
+    if Δgauge > tol && alg_rrule.verbosity >= 1
+        @warn "`eigsolve` cotangents sensitive to gauge choice: (|Δgauge| = $Δgauge)"
+    end
```

Comment:
That makes perfect sense. I am wondering though what you think about only warning for this when verbosity==1. I have to admit I kind of changed it like this, because the current tests, which compute the full Jacobian, were producing to many of those warnings. In actual use cases, I would expect that the warning could be useful to indicate possible unintended behaviour and should maybe not be hidden behind the verbosity flag.